docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

```diff
@@ -121,6 +124,176 @@ network usage.
 Check out the formal specification
 [here](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md).
 
+### Implementation
+
+There are two primary modes of usage right now:
+
+1) Trusted RPC proxy (wrapping an RPC client + verification)
+2) Part of the IBC light client (only verification bit, no RPC) [spec](https://github.com/cosmos/ics/tree/master/spec/ics-002-client-semantics)
+
+First, we'll need something, which will provide us secure headers & validator sets.
+
+```go
+type Provider interface {
+	// 0 - latest
+	SignedHeader(height int64) (SignedHeader, error)
+
+  // 0 - latest
+  Validators(height int64) (ValidatorSet, error)
+}
+```
+
+In case of the proxy it will be a `http` provider (wrapping RPC client). For
+IBC, it will be a `ibc` provider, receiving information from IBC transactions.
+
+Once we have the information, we need to verify it.
+
+```go
+type mode int
+
+const (
+	sequential mode = iota
+	bisecting
+)
+
+// default mode - DefaultBisectingVerification
+type Verifier struct {
+	chainID            string
+	options            TrustOptions
+	lastVerifiedHeight int64
+	logger             log.Logger
+
+	mode       mode
+	trustLevel float
+
+	// Source of new headers.
+	source Provider
```

Comment:
Don't think the `source` or `verifiers` make sense in the context of IBC?

Reply:
relayer can be a source. I will rename `verifiers` to `alternativeSources`, which in IBC case, can be other relayers.

Reply:
The expected behavior during IBC is that a relayer will perform the bisection and submit all the headers downloaded during the bisection in the IBC packet transaction.

The light client verifier on chain should then sequentially validate the headers provided.